### PR TITLE
 CA-292908 (fixed progress bar) + CA-293153 (corrected unwind precheck actions) + improvements

### PR DESCRIPTION
**Pull request by kc284**

### CODE REVIEW

File: XenAdmin/Wizards/PatchingWizard/AutomatedUpdatesBasePage.cs [hunk 1]
```diff
@@ -201,20 +208,24 @@ private void WorkerProgressChanged(object sender, ProgressChangedEventArgs e)
...
-                UpdateStatusTextBox();
+                UpdateStatus();
             }
         }
 
-        private void UpdateStatusTextBox()
+        private void UpdateStatus()
         {
+            // only increase the progress if the action succeeded
+            var newVal = backgroundWorkers.Sum(b => b.PercentComplete) / backgroundWorkers.Count;
```
Comment (MihaelaStoica) [inline XenAdmin/Wizards/PatchingWizard/AutomatedUpdatesBasePage.cs]: When the page is cancelled, this method is called from WorkerCompleted (which is triggered by the cancellation), but `backgroundWorkers.Count` is 0, due to `backgroundWorkers.Clear()` in PageCancelled(), so the progressBar value cannot be calculated in this case.

File: XenAdmin/Wizards/PatchingWizard/PatchingWizard_PrecheckPage.cs [hunk 2]
```diff
@@ -642,9 +642,23 @@ public override bool EnableNext()
         public Pool_patch Patch { private get; set; }
         public Pool_update PoolUpdate { private get; set; }
 
+        public List<AsyncAction> GetUnwindChangesActions()
```
Comment (MihaelaStoica) [inline XenAdmin/Wizards/PatchingWizard/PatchingWizard_PrecheckPage.cs]: Everytime this function is called (when moving away from the prechecks page and also on cancelling the wizard), it creates the unwind actions and the actions that are created on leaving the page are shown in the Events page before they are actually executed. There will also be duplicates when moving back to the prechecks page and then to the next page again (this is possible in the RPU wizard only)

File: XenAdmin/Wizards/PatchingWizard/PatchingWizard.cs [hunk 3]
```diff
@@ -296,8 +284,11 @@ protected override void OnCancel()
         {
             base.OnCancel();
 
-            var subActions = BuildSubActions(PatchingWizard_PrecheckPage.GetUnwindChangesActions,
-                GetRemovePatchActions, GetRemoveVdiActions, GetCleanUpPoolUpdateActions);
+            var subActions = new List<AsyncAction>();
+            subActions.AddRange(GetUnwindChangesActions(PatchingWizard_PrecheckPage.PrecheckProblemsActuallyResolved));
```
Comment (MihaelaStoica) [inline XenAdmin/Wizards/PatchingWizard/PatchingWizard.cs]: AddRange fails with ArgumentNullException if any of these functions (GetUnwindChangesActions, GetRemovePatchActions, etc.) returns null

Comment (kc284) [reply XenAdmin/Wizards/PatchingWizard/PatchingWizard.cs]: oops! 😳

**Pull request merged**